lib/graphql/schema/filter.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module GraphQL
+  class Schema
+    class Filter
+      def initialize(field:)
+        @field = field
+      end
+
+      def resolve_field(obj, args, ctx)
+        yield(obj, args, ctx)
+      end
+
+      private
+
+      # TODO rename, use elsewhere?
+      # @api private
+      class LazyThingy
+        def initialize(schema, lazy_value, &then_block)
+          @schema = schema
+          @lazy_value = lazy_value
+          @then_block = then_block
+        end
+
+        def value
+          if defined?(@value)
+            @value
+          else
+            method = @schema.lazy_method_name(@lazy_value)
+            next_value = @lazy_value.public_send(method)
+            @value = @then_block.call(next_value)
+          end
+        end
+      end
+
+      def after_lazy(schema, value)
+        if schema.lazy?(value)
+          LazyThingy.new(schema, value, &Proc.new)
+        else
+          yield(value)
+        end
+      end
+    end
+
+    class ConnectionFilter < Filter
+      def initialize(field:)
+        field.argument :after, "String", "Returns the elements in the list that come after the specified global ID.", required: false
+        field.argument :before, "String", "Returns the elements in the list that come before the specified global ID.", required: false
+        field.argument :first, "Int", "Returns the first _n_ elements from the list.", required: false
+        field.argument :last, "Int", "Returns the last _n_ elements from the list.", required: false
+        super
+      end
+
+      def resolve_field(obj, args)
+        inner_args = args.dup
+        inner_args.delete(:first)
+        inner_args.delete(:after)
+        inner_args.delete(:last)
+        inner_args.delete(:before)
+        nodes = yield(obj, inner_args)
+        after_lazy(obj.context.schema, nodes) do |value|
+          if value.nil?
+            nil
+          elsif value.is_a?(GraphQL::Execution::Execute::Skip)
```

Comment:
It would be great if the filter base class could handle these cases somehow (and the lazy resolution case). In our filters we ended up with two methods a la before/after instead of a single `yield`able resolve, so that the base class could do this logic in the middle. That said I do really prefer the `yield` pattern because it's easy to understand and more in line with how middleware works everywhere else...

Reply:
I'm needing this same thing in another branch, and trying out `Schema#after_lazy`: 

https://github.com/rmosolgo/graphql-ruby/blob/9ae8e86e5ff15de740e213288f42162b61014ba8/lib/graphql/schema.rb#L968-L981

It's a bit odd to have it as a method on `schema`, but it's because each schema maintains a map of `class => method` pairs for resolution. 

There's a crazy (🙈 ) easier-to-use solution which would involve adding `.graphql_lazy?`, `.graphql_then(&:block)` and `.graphql_sync` to `Object`, and then letting people re-implement those methods on should-be lazy objects (eg, `Promise`). So you could always use `obj.graphql_then { |resolved_obj| ... }` , and for most objects, it would just be `yield_self`. 

I think I'll try it on another branch sometime. 

(Funny, it's technically a feature where refinements could work, but I think we support some Rubies that never implemented `refine`.)